lib/galaxy/workflow/run_request.py

```diff
@@ -265,6 +265,17 @@ def build_workflow_run_configs(trans, workflow, payload):
             # inputs with referential integrity if parameters are already normalized (coming from tool form).
             normalized_inputs = {}
 
+        if param_map:
+            # disentangle raw parameter dictionaries into formal request structures if we can
+            # to setup proper WorkflowRequestToInputDatasetAssociation, WorkflowRequestToInputDatasetCollectionAssociation
+            # and WorkflowRequestInputStepParameter objects.
+            for step in workflow.steps:
+                normalized_key = step.id
+                if step.type == "parameter_input":
```

Comment:
This still trips up when there are `data_input` parameters in the workflow:
```
galaxy.workflow.run ERROR 2018-11-05 17:28:53,706 Failed to schedule Workflow[id=287,name=NonDataInputModule], problem occurred on WorkflowStep[index=1,type=data_input].
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/run.py", line 190, in invoke
    incomplete_or_none = self._invoke_step(workflow_invocation_step)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/run.py", line 266, in _invoke_step
    use_cached_job=self.workflow_invocation.use_cached_job)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/modules.py", line 535, in execute
    progress.set_outputs_for_input(invocation_step, step_outputs)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/run.py", line 393, in set_outputs_for_input
    raise ValueError(message)
ValueError: Step with id 1008 not found in inputs_step_id ({1007: 'new label'})
galaxy.workflow.run ERROR 2018-11-05 17:28:53,708 Failed to execute scheduled workflow.
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/run.py", line 83, in __invoke
    outputs = invoker.invoke()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/run.py", line 190, in invoke
    incomplete_or_none = self._invoke_step(workflow_invocation_step)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/run.py", line 266, in _invoke_step
    use_cached_job=self.workflow_invocation.use_cached_job)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/modules.py", line 535, in execute
    progress.set_outputs_for_input(invocation_step, step_outputs)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/workflow/run.py", line 393, in set_outputs_for_input
    raise ValueError(message)
ValueError: Step with id 1008 not found in inputs_step_id ({1007: 'new label'})
```

Reply:
```
diff --git a/lib/galaxy/workflow/run_request.py b/lib/galaxy/workflow/run_request.py
index ff49501943..1dd955dd9d 100644
--- a/lib/galaxy/workflow/run_request.py
+++ b/lib/galaxy/workflow/run_request.py
@@ -271,10 +271,13 @@ def build_workflow_run_configs(trans, workflow, payload):
             # and WorkflowRequestInputStepParameter objects.
             for step in workflow.steps:
                 normalized_key = step.id
-                if step.type == "parameter_input":
+                if step.type in ["parameter_input", 'data_input']:
                     if normalized_key in param_map:
                         value = param_map.pop(normalized_key)
-                        normalized_inputs[normalized_key] = value["input"]
+                        if step.type == 'data_input':
+                            normalized_inputs[normalized_key] = value["input"]['values'][0]
+                        else:
+                            normalized_inputs[normalized_key] = value["input"]

         steps_by_id = workflow.steps_by_id
         # Set workflow inputs.
```

this let's me run the workflow -- but I'm guessing this will fall flat in many scenarios :/

Reply:
But the tests were passing and we weren't touching those inputs at all in this loop before, I don't get it. Would it only fail on workflows that had both data and parameter inputs?

Reply:
I'll write some tests and try to verify this is needed.

Reply:
Okay - I think the problem is here (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/workflow/run.py#L393). inputs_by_id is now populated by only partially, but if outputs is supplied and has an 'output' value we can skip the if for that step. I can't make the change tonight, but I'll do it in the morning. I think this is a lighter touch than the PR - is that okay?

Reply:
Yeah, that's more than reasonable, I did play around this but didn't realize we could just skip if the output is there ... which makes a lot of sense.